macros/Compat.ejs

```diff
@@ -1,66 +1,399 @@
+
+*/
+function getVersionString(versionInfo) {
+  switch (versionInfo) {
+    case null:
+      return `<span title="${localize(compatStrings, 'supportsShort_unknown_title')}"
+              style="color: rgb(255, 153, 0);">${localize(compatStrings, 'supportsShort_unknown')}</span>`;
+      break;
+    case true:
+      return `<span style="color: #888">(${localize(compatStrings, 'supportsShort_yes')})</span>`;
+      break;
+    case false:
+      return `<span style="color: #f00">(${localize(compatStrings, 'supportsShort_no')})</span>`;
+      break;
+    default:
+      return versionInfo;
+  }
+}
+
+/*
+  Create the note for a browser flag or preference
+*/
+function generateFlagsNote(supportData, browserId) {
+  var versions = '';
+  if (typeof(supportData.version_added) === 'string') {
+    versions = 'From version ' + supportData.version_added;
+  }
+  if (typeof(supportData.version_removed) === 'string') {
+    if (versions) {
+      versions += ' until version '+ supportData.version_removed + ' (excluded), ';
     } else {
-      output = template("CompatDataTable", [data, "module"]);
+      versions = 'Until version ' + supportData.version_removed + ' (excluded), ';
     }
+  } else {
+    versions += ', ';
+  } if ((browserId === 'firefox') || (browserId === 'firefox_android')) {
+    if (supportData.flag.type === 'preference') {
+      action = 'this feature ' + ((typeof(supportData.version_removed) === 'string')?'was':'is') + ' available behind the <i>about:config</i> preference <code>' + supportData.flag.name + '</code> (Activated value: ' + supportData.flag.value_to_set + '</code>).';
+    } else {
+      action = 'this feature ' + ((typeof(supportData.version_removed) === 'string')?'was':'is') + 'behind the compile flag <code>' + supportData.flag.name + '</code> (value: <code>' + supportData.flag.value_to_set + ')</code>.';
+    }
+  } else if ((browserId === 'chrome') || (browserId === 'chrome_android')) {
+    if (supportData.flag.type === 'preference') {
+      action = 'this feature '+ ((typeof(supportData.version_removed) === 'string')?'was':'is') + ' available behind the <i>' + supportData.flag.name + '</i> flag in <code>chrome://flags</code>.';
+    }
+  } else {
+    action = 'this feature ' + ((typeof(supportData.version_removed) === 'string')?'was':'is') + ' available behind the <code>' + supportData.flag.name + '</code> flag to <code>' + supportData.flag.value_to_set + '</code>.';
+  }
+  return versions + action;
 }
 
-if (output === '') {
-  output = noDataFoundl10n;
+
+/*
+  Do all the magic needed to get the content of a support cell in the table
+*/
+function getSupportInfo(supportData, browser, compatNotes) {
+  let supportInfo = '';
+
+  // browsers are optional in the data, display them as "?" in our table
+  if (!supportData) {
+    supportInfo = getVersionString(null);
+  // we have support data, lets go
+  } else {
+    // Handling of first version added
+    if (supportData.version_added === undefined) {
```

Comment:
You are absolutely right, I copy-pasted too much here and indeed the `getVersionString` function should just handle this. I've removed this duplication now.